compiler/optimizations/remoteValueForwarding.cpp

```diff
@@ -237,18 +256,21 @@ static bool canForwardValue(Map<Symbol*, Vec<SymExpr*>*>& defMap,
   } else if (isAtomicType(arg->type)) {
     retval = false;
   } else if (arg->intent == INTENT_CONST_REF) {
-    retval = true;
+    DotInfo* info = dotLocaleMap[arg];
+    if (info == NULL) {
+      retval = true;
+    } else {
+      retval = !info->usesDotLocale;
+    }
   } else if (arg->intent == INTENT_CONST_IN &&
       !arg->type->symbol->hasFlag(FLAG_REF)) {
     // BHARSH TODO: This can currently happen when the arg is the lhs of a +=,
     // but it obviously needs to have the 'ref' intent. One example can be seen
     // for += between strings.
     retval = true;
 
-  } else if (arg->isRef()  == true &&
-             isSafeToDerefArg(defMap, useMap, arg) == true) {
-    retval = true;
-
+  } else if (arg->isRef()  == true) {
+    retval = false;
```

Comment:
I think updateLoopBodyClasses still uses the remaining functions. I'd like to look at that function in more detail some other time to see what could be improved.